### ISSUE TITLE
Ad Collection Issues in Firefox

### ISSUE THREAD

**Issue by HRussellZFAC023** [dhowe/AdNauseam#1364]
## Describe the issue
Many firefox users have been complaining that their ads rarely get clicked - those that do are usually text ads from google searches. I was interested in what was causing this bug.

I decided to play around with the filter lists and noticed that when I disbaled "easylist", ads would then be clicked on (which previously would have been ignored).

Upon further examination and playing with all the settings, I discovered that malware protection setting was causing the conflict - the malware protection (in firefox at least) prevents ad urls in the "easylist" filter from being followed with a block screen.

(As a further note to firefox users, they should disbale tracking protection in settings as this blocks some ads.)

##Example URL:
http://parkermaths.com/index.html
http://www.speedtest.net/

### Screenshot in which the issue can be seen
Malware blocking setting:
![image](https://user-images.githubusercontent.com/22746105/40589339-8951b356-61e3-11e8-9b30-db018d5d6c53.png)
When enabled, this setting blocks the user from visiting sites with ads with a screen like this:
![image](https://user-images.githubusercontent.com/22746105/40589326-65b725ca-61e3-11e8-9564-f1543c341343.png)
That same screen is preventing AdNauseam from working - the software trys to follow the link but can't due to "malware protection"

### Steps for anyone to reproduce the issue

Step 1: install firefox and extension
Step 2: go to any website with ads
Step 3: disable adnausium for that site (big blue power Ublock button)
Step 4: refresh - you should now see ads
Step 5: click on an add and get the blocked screen

After doing this, you can disable malware protection and repeat steps 1-4 and notice you will not be blocked.
Renable the extension for that page (blue button again) and then you will also notice the ads will be visited! 

### Your settings

- OS/version:windows 10
- Browser/version:firefox 60
- AdNauseam version:3.5.1
- Other extensions you have installed:none

### Your location/country
(If the issue is website specific, knowing your approximate location will help us to reproduce it)
UK

**Comment by cqx931** [comment]
Hi @HRussell63250, thank you for your time and effort in this issue!
I have tried to recreate it but failed with Step5. If I click on an ad, the corresponding web page shows up as expected and there is no blocked screen for me.

OS/version:macOS Sierra 10.12.6
Browser/version:firefox 60
AdNauseam version:3.5.1
Other extensions you have installed:none

**Comment by HRussell63250** [comment]
Yes you are correct, I have tried this on more ads now. The blocked screen on will appear if you right click on an add and "press open link in a new tab", however if you click on it normally it redirect as expected. 

Can you try to replicate on the below ad link? 

https://adclick.g.doubleclick.net/aclk?sa=l&ai=CmxMep9wLW7yTH4qOtge9hqH4DILJzOtQ1tSg9-QFwI23ARABIOfygiZgu4aAgNAKoAHPnr_WA8gBCakCcW2fD9xdtj6oAwHIA8sEqgSgAU_Qjo1PSvqHxwjeq19Bl4TD43G9RZ-6TOdx2pvVFkOWvagxRzutK6F6tj3W7S7M0Y8bp2bWhq5hh86aghMLhwMF3cfzpsBDunBptsA12krNLN0TGrws_23mhas4NSZVAITJxncD_8j4L9OB6jeoqxpeY9OhWCyt_m3VNQO5rOnBxaOwLt_18axuEFh2OecrcjWTUjZFxhJXYCgMxwyi9wqgBi6AB5nhwCmoB47OG6gHjc0bqAfVyRuoB6a-G6gHmM4b2AcA0ggJCIDhgHAQARgCsQlo8NcuF0S5aYAKAdgTDYgUAg&num=1&sig=AOD64_1C3SvANnFVD6Q7iIrsvw4-dA8pmA&client=ca-pub-9451489399289911&adurl=https://www.jca-adventure.co.uk/offers&nx=448&ny=60&mb=2&bg=!pKelp79EFzI4aMRxj6ICAAAAUVIAAAAXCgADYuvImQEmOGVJZabgvvRZzyan0ohvh8GSoVgcXrFyzGy-eME2Q_REl2dnOY6wLeJaAaIBvcbVihUOC9cwkTGYtnCuUy4Jn1cvdqhGT7CzGXThfisOn5aMKSEAn5vOO-1vxqE-aOYkxMq2RwfmCFW-2rITqIUI6LuMJkLlDffcCBUmfAiE5017TAywLzUSUCD5xj6LYQQnTjFsJ9XFozkUDQJTi33DHNTF5LHZcPjizZcIxbatAiOrYmUkov2ufcFXycY0t5H7O-sQ4F3HcfLVwUvBIDn2Lq26CmOl9tMJ23m83VGhk9wHLrg68xGfZeBs3BV0E2PFtAoe_6a3DnjpxET1Ed5FSa7QnW0Zfyb6ytIIrbjAqUfmf2rbO_W5ocDG_npcolHZ-2wrz6ao

**Comment by cqx931** [comment]
>  The blocked screen on will appear if you right click on an add and "press open link in a new tab"

This is the correct behavior we inherit from uBlock origin, and it is the same in chrome based browsers.

However, I still can't reproduce the block screen with the above link or with opening in a new tab, which is really strange...

But it is really interesting that unselecting the Malware/Blocking function would make AdNauseam in Firefox clicking more ads. If you would like to spend more time into this issue, I would suggest you checking the debugging option in settings and look into the logger(open the menu -> click the 'u' button on the bottom left corner -> click the first icon above 'requests blocked')
https://github.com/gorhill/uBlock/wiki/The-logger
The logger works almost in the same way as the uBlock logger with a few more colors that are unique to adnauseam.

**Comment by HRussell63250** [comment]
This shows what happens when an add is right clicked and opened in a new tab:
![2018-05-28_12-25-37](https://user-images.githubusercontent.com/22746105/40612454-7df2aeea-6272-11e8-9adb-d9133af687b1.gif)
Some ads however will be immediately blocked if clicked:
![2018-05-28_12-30-32](https://user-images.githubusercontent.com/22746105/40612597-1dafa406-6273-11e8-9ff0-db974d04f4d4.gif)
When the above scenario occurs, the logger looks like this

**Comment by HRussell63250** [comment]
![image](https://user-images.githubusercontent.com/22746105/40612779-cb58cb64-6273-11e8-95a7-b9dcec5d69f3.png)

**Comment by Dyras** [comment]
Not sure if this is the right place, but I can't get AdNauseum to block ads at all in Firefox/Basilisk/Waterfox :(

![image](https://user-images.githubusercontent.com/12563436/42049315-e3659afc-7b05-11e8-8073-7f733c3846aa.png)

**Comment by dhowe** [comment]
@cqx931 what is the status of this ticket?

**Comment by dhowe** [comment]
@cqx931 what is the status of this ticket?

**Comment by cqx931** [comment]
Not sure whether this could be an windows issue or not, but from my experience with Firefox, I havn't experienced what when ads are rarely get clicked. And the blocking screen should not be related to the clicking in background.
The current blocking screen when 'opening an ad in new tab' consistent should be ok as it is supposed to block any advertisements popping up automatically.

**Comment by dhowe** [comment]
@cqx931  - Next steps?

**Comment by cqx931** [comment]
One interesting I found by visiting https://www.speedtest.net/ both in chrome and firefox is that only in firefox, there is one rule in easylist that should be allowed by adnauseam but is somehow getting blocked right now. Once I enable this rule with a dynamic filter, all the ads are get rendered successfully in firefox, which explains what @HRussellZFAC023 found out in this ticket.

![image](https://user-images.githubusercontent.com/3056154/64269333-b3d03c00-cf07-11e9-8f4e-029803aa7cbc.png)

Same for http://parkermaths.com/index.html. Some other rules from easylist are blocked 
![image](https://user-images.githubusercontent.com/3056154/64269638-3f49cd00-cf08-11e9-8a8b-1533369d3c4f.png)


However, there are rules from easylist that are correctly allowed by adnauseam in firefox. Going to look into the issue and see why some rules from easylist are blocked by adnauseam in firefox.

**Comment by cqx931** [comment]
I can't reproduce this consistently. I could reproduce this once, but now I am loosing track of it again. At least what described above should be related to any issues regarding to 0 ads collected in Firefox.

**Comment by dhowe** [comment]
How to proceed?

**Comment by cqx931** [comment]
Two places to check: 
1. isBlockableRequest()
https://github.com/dhowe/AdNauseam/blob/master/src/js/adn/core.js#L1139
This is where AdNauseam decide whether to block a request or not. Although I don't think the issue lies here...

2.Pagestore -> filterRequest()
It's more likely that something might goes wrong here in Pagestore.This is the place where AdNauseam's standard of blocking is checked. However, what's the current difference in request handling in Chrome and Firefox? Why does this issue only occurs in Firefox and not in Chrome?
https://github.com/dhowe/AdNauseam/blob/master/src/js/pagestore.js#L645

**Comment by dhowe** [comment]
@cqx931 this is a critically important issue, but it is a bit difficult to follow above. Can you provide a clear description (either here or in a fresh ticket) and include a reproducible test case? Thanks

**Comment by valuesindesign** [comment]
Okay, mystery fixed. This is way worse than I expected. And the bug has been there for ages...
It's actually a bug in list updating with adnauseam. Fixed here: https://github.com/dhowe/AdNauseam/commit/edf8941528629ad4bed7e75ea75b1e343c3d7cc3

The reason why these easylist rules are blocked is because these are 'new' rules that don't exist in the easylist version that comes with adnauseam. In each list update, the lists are correctly updated in uBlock origin's mechanism, but not in adnauseam. Therefore, when a 'new' blocking rule is being sent to adnauseam and ask whether this should be allowed, adnauseam can't find it in existing rules, which will result in a return of 'block' instead of  'allow'(for easylist).

It's also not a Firefox only issue. The reason why it has different behavior in firefox and chrome could potentially related to how we build and test in chrome and firefox. In chrome, whenever I test adnauseam from a new build, the lists used are based on whatever comes prepacked. Unless I manually update everything. However, in Firefox, if I already have adnauseam installed and run the firefox script, it's going to use the list data from my usual profile(which has been updated), therefore resulting a conflict of what I described above. 

I will work on an emergent release update after this.
In

**Comment by cqx931** [comment]
Okay, mystery fixed. This is way worse than I expected. And the bug has been there for ages...
It's actually a bug in list updating with adnauseam. Fixed here: https://github.com/dhowe/AdNauseam/commit/edf8941528629ad4bed7e75ea75b1e343c3d7cc3

The reason why these easylist rules are blocked is because these are 'new' rules that don't exist in the easylist version that comes with adnauseam. In each list update, the lists are correctly updated in uBlock origin's mechanism, but not in adnauseam. Therefore, when a 'new' blocking rule is being sent to adnauseam and ask whether this should be allowed, adnauseam can't find it in existing rules, which will result in a return of 'block' instead of  'allow'(for easylist).

It's also not a Firefox only issue. The reason why it has different behavior in firefox and chrome could potentially related to how we build and test in chrome and firefox. In chrome, whenever I test adnauseam from a new build, the lists used are based on whatever comes prepacked. Unless I manually update everything. However, in Firefox, if I already have adnauseam installed and run the firefox script, it's going to use the list data from my usual profile(which has been updated), therefore resulting a conflict of what I described above. 

I will work on an emergent release update after this.
